proposals/stdlib/mutablelist-move.md

```diff
@@ -0,0 +1,145 @@
+# Adding moveXXX() Functions to MutableList
+
+It would be helpful to create some additional extension functions for `MutableList` to move items to new indexes, as well as move them up/down, and move one or all items that meet a predicate.
+
+
+# Title
+
+* **Type**: Standard Library API proposal
+* **Author**: Thomas Nield
+* **Contributors**: N/A
+* **Status**: Submitted
+* **Prototype**: Implemented
+
+
+## Summary
+
+It would be helpful to create some additional extension functions for `MutableList` to move items to new indexes, as well as move them up/down, and one or all items that meet a predicate.
+
+These can be implemented using only extension functions on `MutableList`.
+
+## Similar API review
+
+Cannot think of an explicit precendent in another API.
+
+## Use cases
+
+* Common use cases often have to do with re-ordering items especially in a UI context, and any scenario dealing with mutable ordered collections may be candidate.
+
+* In JavaFX and other UI technologieis, a user may want to move an item in a `MutableList` up or down to define some arbitrary order.
+
+* Items can be searched in a `MutableList` by having the matching items moved to the top of the `MutableList`, rather than creating a new `List` or some filtering abstraction
+
+* Observe this [demo application](https://github.com/thomasnield/rxkotlinfx-tornadofx-demo) to see several of these use cases in action.
+
+## Alternatives
+
+Developers could implement the extension functions themselves, as shown below. Moving items can be tricky and verbose to coordinate with local implementations.
+
+```kotlin
+/**
+ * Moves the given **T** item to the specified index
+ */
+fun <T> MutableList<T>.move(item: T, newIndex: Int)  {
+    val currentIndex = indexOf(item)
+    if (currentIndex < 0) return
+    removeAt(currentIndex)
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given item at the `oldIndex` to the `newIndex`
+ */
+fun <T> MutableList<T>.move(oldIndex: Int, newIndex: Int)  {
+    val item = this[oldIndex]
+    removeAt(oldIndex)
+    add(newIndex, item)
```

Comment:
I may need to rename this to `moveAt()` or something similar so the overload doesn't clash if its a `MutableList<Int>`. Is that what you were thinking with previous comment?

Good catch on inconsistent state. It crossed my mind but I reasoned incorrectly as my thought experiments thought of the item only moving up, not down. I may need to do a check to see if an offset occurs and adjust by `1` accordingly.